File: cli-rs/crates/alp-core/src/wizard/filesystem.rs
// SPDX-License-Identifier: Apache-2.0
//! Filesystem I/O for the wizard — diff detection and writing.

use std::path::Path;

use super::models::{WizardFileChange, WizardFileChangeKind, WizardPlannedFile, WizardWriteResult};

/// Classify each planned file as New / Update / Unchanged relative to `project_root`.
pub fn collect_wizard_file_changes(
    project_root: &Path,
    files: &[WizardPlannedFile],
) -> Vec<WizardFileChange> {
    files
        .iter()
        .map(|f| {
            let path = project_root.join(&f.relative_path);
            let kind = if !path.exists() {
                WizardFileChangeKind::New
            } else {
                match std::fs::read_to_string(&path) {
                    Ok(existing) if existing == f.content => WizardFileChangeKind::Unchanged,
                    _ => WizardFileChangeKind::Update,
                }
            };
            WizardFileChange { relative_path: f.relative_path.clone(), kind }
        })
        .collect()
}

/// Write all planned files under `project_root`, creating parent directories as needed.
/// Files whose content is already identical are skipped (counted as unchanged).
pub fn write_wizard_files(
    project_root: &Path,
    files: &[WizardPlannedFile],
) -> Result<WizardWriteResult, std::io::Error> {
    let mut written = Vec::new();
    let mut unchanged = Vec::new();

    for f in files {
        let path = project_root.join(&f.relative_path);

        if path.exists() {
            if let Ok(existing) = std::fs::read_to_string(&path) {
                if existing == f.content {
                    unchanged.push(f.relative_path.clone());
                    continue;
                }
            }
        }

        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        std::fs::write(&path, &f.content)?;
        written.push(f.relative_path.clone());
    }

    Ok(WizardWriteResult { written, unchanged })
}

File: packages/alp-core/src/wizard/fileSystem.ts
// SPDX-License-Identifier: Apache-2.0

import * as fs from "fs";
import * as path from "path";
import { listGenerationTargetSupport } from "../loader/service";
import {
  WizardFileChange,
  WizardGeneratedOutputPreview,
  WizardPlannedFile,
  WizardWriteResult,
} from "./models";

export function collectWizardFileChanges(
  workspaceRoot: string,
  files: WizardPlannedFile[],
): WizardFileChange[] {
  return files.map((file) => {
    const absolutePath = path.join(workspaceRoot, file.relativePath);
    if (!fs.existsSync(absolutePath)) {
      return {
        relativePath: file.relativePath,
        kind: "new",
      };
    }

    const current = fs.readFileSync(absolutePath, "utf-8");
    return {
      relativePath: file.relativePath,
      kind: current === file.content ? "unchanged" : "update",
    };
  });
}

export function writeWizardFiles(
  workspaceRoot: string,
  files: WizardPlannedFile[],
): WizardWriteResult {
  const result: WizardWriteResult = {
    written: [],
    unchanged: [],
  };

  for (const file of files) {
    const absolutePath = path.join(workspaceRoot, file.relativePath);
    const exists = fs.existsSync(absolutePath);
    if (exists) {
      const current = fs.readFileSync(absolutePath, "utf-8");
      if (current === file.content) {
        result.unchanged.push(file.relativePath);
        continue;
      }
    }

    fs.mkdirSync(path.dirname(absolutePath), { recursive: true });
    fs.writeFileSync(absolutePath, file.content, "utf-8");
    result.written.push(file.relativePath);
  }

  return result;
}

export function collectGeneratedOutputPreviews(
  workspaceRoot: string,
): WizardGeneratedOutputPreview[] {
  const targets = listGenerationTargetSupport();

  return targets.map((target) => {
    const absolutePath = path.join(workspaceRoot, target.outputRelativePath);
    if (!fs.existsSync(absolutePath)) {
      return {
        emit: target.emit,
        displayName: target.displayName,
        outputRelativePath: target.outputRelativePath,
        languageId: target.preview.languageId,
        state: "missing",
        contentPreview:
          "(Not generated yet. Run Alp: Generate all to create this output.)",
      };
    }

    const content = fs.readFileSync(absolutePath, "utf-8");
    const preview = content.split(/\r?\n/).slice(0, 18).join("\n").trim();
    return {
      emit: target.emit,
      displayName: target.displayName,
      outputRelativePath: target.outputRelativePath,
      languageId: target.preview.languageId,
      state: "existing",
      contentPreview:
        preview.length > 0 ? preview : "(File exists but is empty.)",
    };
  });
}
